feat: Support extensional equality in FinFun

fmontesi opened the pull request:
This PR changes `FinFun` to support extensional equality, making it essentially equivalent to `Finsupp` in Mathlib, with the important difference that definitions and theories about `FinFun` are computable. It also makes FinFun adopt the same notation as Finsupp, so to facilitate a future replacement with the latter in the future if it ever changes in Mathlib to have computable defs/theorems (this is relegated to the future since FinFun is used downstream in at least one big project I know of).

chenson2018 commented on Cslib/Foundations/Data/FinFun.lean:
```diff
@@ -6,108 +6,106 @@ Authors: Fabrizio Montesi, Xueying Qin
+@[scoped grind =]
+theorem coe_eq_fn [Zero β] {f : α →₀ β} : f a = f.fn a := by
+  rcases f with ⟨f, support, mem_support_fn⟩
+  simp [DFunLike.coe]
```
```suggestion
  simp [DFunLike.coe]
```

chenson2018 commented on Cslib/Foundations/Data/FinFun.lean:
```diff
@@ -6,108 +6,106 @@ Authors: Fabrizio Montesi, Xueying Qin
+    intro x
+    rw [coe_fn]
+    rw [coe_fn]
+    rw [h1]
```
```suggestion
theorem eq_char [DecidableEq α] [Zero β] {f g : α →₀ β} :
    f = g ↔ (f.fn = g.fn ∧ f.support = g.support) := by
  have (h : f.fn = g.fn) : f = g := ext (congrFun h)
  grind
```

chenson2018 replied:
The reverse direction works without the support hypothesis, this is intentional?

XYUnknown commented on Cslib/Foundations/Data/FinFun.lean:
```diff
@@ -6,108 +6,106 @@ Authors: Fabrizio Montesi, Xueying Qin
-def FinFun.defined (f : α ⇀ β) (x : α) : Prop := x ∈ f.dom
-
-@[simp]
-abbrev FinFun.apply (f : α ⇀ β) (x : α) : β := f.f x
```
It would be convenient if we keep this

fmontesi replied:
Why? You can now just write `f x` and it'll do the same thing, because of the new `FunLike` instance. Making that an `abbrev` reveals the internal details of `FinFun` a bit.

XYUnknown reviewed (commented):
It should be compatible with my code base. 
Two small syntactic things: it would be convenient if we keep the abbreviation `FinFun.apply` and the definition`Function.toFinFun`.

fmontesi commented on Cslib/Foundations/Data/FinFun.lean:
```diff
@@ -6,108 +6,106 @@ Authors: Fabrizio Montesi, Xueying Qin
+    have hmem := f.mem_support_fn (a := a)
+    simp only [Membership.mem] at hmem
+    simp only [Membership.mem]
+    grind
```
Looks like I forgot to backtrack and transform some proofs into `grind` calls. Gonna call it `grind` debt. ;-)

fmontesi commented on Cslib/Foundations/Data/FinFun.lean:
```diff
@@ -6,108 +6,106 @@ Authors: Fabrizio Montesi, Xueying Qin
-    : (FinFun.mapBin f g op).isSome := by grind [mapBin]
-
--- Fun to FinFun
-def Function.toFinFun [DecidableEq α] (f : α → β) (dom : Finset α) : α ⇀ β := FinFun.mk f dom
```
This is now `FinFun.mkRestrictFun`, accessible through the uprighthook notation: `f ↾ support`.

I made `mkRestrictFun` private for now because I'm not in particular love with the name...... :-)

chenson2018 commented on Cslib/Foundations/Data/FinFun.lean:
```diff
@@ -6,108 +6,90 @@ Authors: Fabrizio Montesi, Xueying Qin
+namespace FinFun
+
+@[inherit_doc]
+scoped infix:25 " →₀ " => FinFun
```
Just making sure this is not a typo: you made this `infix` versus `Finsupp` using `infixr`?

fmontesi commented:
@XYUnknown I've added some results about restriction, which has pretty nice properties, and renamed it to `fromFun`.

@XYUnknown @chenson2018 I am though in doubt whether I should use the `↾` symbol for the 'restriction to a finite support' operator. In general, that operator might make sense for any Set. Maybe I should rename it to `↾₀` ?

XYUnknown commented:
> @XYUnknown I've added some results about restriction, which has pretty nice properties, and renamed it to `fromFun`.

It is indeed a more reasonable name compared to `mkRestrictFun`

> @XYUnknown @chenson2018 I am though in doubt whether I should use the `↾` symbol for the 'restriction to a finite support' operator. In general, that operator might make sense for any Set. Maybe I should rename it to `↾₀` ?

Since the symbol of the `FinFun` is `→₀`, it does make sense to me to use `↾₀`.
